modules/api/src/main/java/org/apache/ignite/lang/IgniteExceptionMapperUtil.java

```diff
@@ -0,0 +1,170 @@
+
+        EXCEPTION_CONVERTERS = unmodifiableMap(mappers);
+    }
+
+    /**
+     * Add a new mapping to already registered ones.
+     *
+     * @param mapper Exception mapper from internal exception to a public one.
+     * @param registeredMappings Already registered mappings.
+     * @throws IgniteException If a mapper for the given {@code clazz} already registered,
+     *      or {@code clazz} represents Java standard exception like {@link NullPointerException}, {@link IllegalArgumentException}.
+     */
+    static void registerMapping(
+            IgniteExceptionMapper<?, ?> mapper,
+            Map<Class<? extends Exception>, IgniteExceptionMapper<?, ?>> registeredMappings) {
+        if (registeredMappings.containsKey(mapper.mappingFrom())) {
+            throw new IgniteException(
+                    INTERNAL_ERR,
+                    "Failed to register exception mapper, duplicate found [class=" + mapper.mappingFrom().getCanonicalName() + ']');
+        }
+
+        if (isJavaStandardException(mapper.mappingFrom())) {
+            throw new IgniteException(
+                    INTERNAL_ERR,
+                    "Failed to register exception mapper. "
+                            + "Mapping for this class is prohibited [class=" + mapper.mappingFrom().getCanonicalName() + ']');
+        }
+
+        registeredMappings.put(mapper.mappingFrom(), mapper);
+    }
+
+    /**
+     * This method provides a mapping from internal exception to Ignite public ones.
+     *
+     * <p>The rules of mapping are the following:</p>
+     * <ul>
+     *     <li>any instance of {@link Error} is returned as is, except {@link AssertionError}
+     *     that will always be mapped to {@link IgniteException} with the {@link ErrorGroups.Common#INTERNAL_ERR} error code.</li>
+     *     <li>any instance of Java standard exception like {@link NullPointerException} is returned as is.</li>
+     *     <li>any instance of {@link IgniteException} or {@link IgniteCheckedException} is returned as is.</li>
+     *     <li>if there are no any mappers that can do a mapping from the given error to a public exception,
+     *     then {@link IgniteException} with the {@link ErrorGroups.Common#INTERNAL_ERR} error code is returned.</li>
+     * </ul>
+     *
+     * @param origin Exception to be mapped.
+     * @return Public exception.
+     */
+    public static Throwable mapToPublicException(Throwable origin) {
+        if (origin instanceof Error) {
+            if (origin instanceof AssertionError) {
+                return new IgniteException(INTERNAL_ERR, origin);
+            }
+
+            return origin;
+        }
+
+        if (origin instanceof IgniteException || origin instanceof IgniteCheckedException) {
+            return origin;
+        }
+
+        if (isJavaStandardException(origin)) {
+            return origin;
+        }
+
+        IgniteExceptionMapper<? extends Exception, ? extends Exception> m = EXCEPTION_CONVERTERS.get(origin.getClass());
```

Comment:
Why do we check only exact matching? Let's say InternalTransactionException -> PublicTransactionException mapper was registered. And we have SomeSpecficInternalTransactionException extends InternalTransactionException. Do we expect this SomeSpecficInternalTransactionException to be converted to PublicTransactionException? I believe we should, because from the hierarchy point of view SomeSpecficInternalTransactionException is an InternalTransactionException. 
Thus I believe we should find the most specific mapper with a top level IgniteException(INTERNAL_ERR, origin);

Reply:
Well, It is arguable. IMHO, a particular component should provide a comprehensive list of mappers. Let's consider the following case:
```
    public class ComponentException {};
    public class SpecificExceptionA extends ComponentException {};

    // @throws SpecificExceptionA If ...
    publicApiCall() throws SpecificExceptionA
```
and after that, we decided to introduce a new type of exception to clarify the behavior in some particular case:
```
    public class SpecificExceptionB extends ComponentException {};
    or
    public class SpecificExceptionB extends SpecificExceptionA{};

    // @throws SpecificExceptionA If ...
    // @throws SpecificExceptionB If ...
    publicApiCall() throws SpecificExceptionA
```
and our component does not provide a mapping from InternalSpecificExceptionB to SpecificExceptionB.

In this case, the user will observe SpecificExceptionA or  ComponentException instead of SpecificExceptionB.
IMHO, this behavior would not be correct.

Reply:
It's correct from the substitution principle. Alternative here, the one you've implemented, is converting SpecificExceptionB to root IgniteException, that should be considered incorrect according to aforementioned logic, meaning we have A extends B extends C, and B to B_public mapper along with anything to C_public. You claim that it's incorrect to consider A as B and thus convert it to B_public however it's correct to consider A as C and thus convert it to C_public, where C_public is IgniteException in this particular case.That seems confusing.
In real life that would actually mean that in case of TransactionInternalCheckedException->TransactionException mapper we will convert LockException extends TransactionInternalCheckedException to IgniteException instead of TransactionException, that is more specific and thus better.

Reply:
Well, I don't think so, just because the logic you proposed breaks handling exceptions on the user's side, and even cannot be compiled in case of checked exceptions.

Let's consider the following examples:
(1)
```
Public hierarchy:
    ComponentException extends IgniteException;
    SpecificExceptionA extends ComponentException
    SpecificExceptionB extends ComponentException

Internal hierarchy:
    InternalComponentException extends IgniteInternalException;
    InternalSpecificExceptionA extends InternalComponentException
    InternalSpecificExceptionB extends InternalComponentException

Mappings:
    InternalComponentException -> ComponentException
    InternalSpecificExceptionA -> SpecificExceptionA
    mapping for InternalSpecificExceptionB is absent due to a bug
```
```
Public api:
    public void apiCall() throws SpecificExceptionA, SpecificExceptionB;

    try {
        apiCall();
    catch (SpecificExceptionB e) {
        // handle specific case (B)
    } catch (SpecificExceptionA e) {
        // handle specific case (A)
    } catch (ComponentException e) {
        // optional catch block, just to illustrate the issue (C)
    }
```

If I understand your proposal correctly, the mapper should convert InternalSpecificExceptionB to ComponentException, because the most specific mapping is InternalComponentException -> ComponentException and InternalSpecificExceptionB extends InternalComponentException.

In this case, exception handler (B) will be skipped at all, only (C) will be executed. I would say, this is a bug caused by the absence of the required mapping and we should clearly state that it is incorrect behavior via throwing `IgniteException(INTERNAL_ERR)` I think.

When ComponentException is a checked exception, the situation is even worse - the code should not be compiled:
```
    public static void apiCall() throws SpecificExceptionA, SpecificExceptionB {
        ...
        throw new ComponentException();
    }
```

(2) 
```
Public hierarchy:
    ComponentException extends IgniteException;
    SpecificExceptionA extends ComponentException
    SpecificExceptionB extends SpecificExceptionA
```

The fact, that the right handler is skipped, is still in play.

Reply:
> I would say, this is a bug caused by the absence of the required mapping and we should clearly state that it is incorrect behavior via throwing IgniteException(INTERNAL_ERR) I think.

Why it's a bug? Is it mandatory to have explicit internal to public mappers for all internal exceptions? Why?

> When ComponentException is a checked exception, the situation is even worse - the code should not be compiled:

What do you mean? You either expect ComponentException to be thrown and thus add it to throws clause, or provide required mapping.